src/bridge/cockpitpcpmetrics.c

```diff
@@ -561,60 +684,24 @@ convert_metric_description (CockpitPcpMetrics *self,
   return TRUE;
 }
 
-static void
-cockpit_pcp_metrics_prepare (CockpitChannel *channel)
+static gboolean
+prepare_current_context (CockpitPcpMetrics *self)
 {
-  CockpitPcpMetrics *self = COCKPIT_PCP_METRICS (channel);
-  const gchar *problem = "protocol-error";
+  gboolean result = FALSE;
   JsonObject *options;
-  const gchar *source;
   gchar **instances = NULL;
   gchar **omit_instances = NULL;
   JsonArray *metrics;
-  const char *name;
-  int type;
   int i;
 
-  COCKPIT_CHANNEL_CLASS (cockpit_pcp_metrics_parent_class)->prepare (channel);
+  g_free (self->metrics);
+  g_free (self->pmidlist);
 
-  options = cockpit_channel_get_options (channel);
+  self->numpmid = 0;
+  self->metrics = NULL;
+  self->pmidlist = NULL;
 
-  /* "source" option */
-  if (!cockpit_json_get_string (options, "source", NULL, &source))
-    {
-      g_warning ("invalid \"source\" option for metrics channel");
-      goto out;
-    }
-  else if (!source)
-    {
-      g_warning ("no \"source\" option specified for metrics channel");
-      goto out;
-    }
-  else if (g_str_equal (source, "direct"))
-    {
-      type = PM_CONTEXT_LOCAL;
-      name = NULL;
-    }
-  else if (g_str_equal (source, "pmcd"))
```

Comment:
We need to file an issue here. pmNewContext() can block for many 10s of seconds if a pmcd is not responsive. This means connecting to pmcd may need to happen in a thread, which has threading implications.